### PR TITLE
[Bug] Careers application finishes if exited

### LINKED ISSUE #51: [Bug] Careers application finishes if exited
when entering an application and click "back to careers" and then go back in to the app to answer the questions, the system thinks the app is done and wont let you do anything

### PR DESCRIPTION
**Pull request by JavaLavaMT**
closes #51

### CODE REVIEW

**Comment by JavaLavaMT** [comment]
image

File: app/controllers/careers/listing/index.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import Controller from '@ember/controller';
+import { inject as service } from '@ember/service';
+import { computed } from '@ember/object';
+
+export default Controller.extend({
+  auth: service(),
+
+  form: computed.reads('forms'),
```
Comment (James1x0) [inline app/controllers/careers/listing/index.js]: Is a reads key here necessary? I'm not sure it is.

File: app/controllers/careers/listing/index.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+  answered: computed(('form', 'applicationStatus'), function () {
+    let formElements = this.form.elements ? this.form.elements.length : 0,
```
Comment (James1x0) [inline app/controllers/careers/listing/index.js]:
```suggestion
    let formElements = (this.form.elements || []).length,
```

File: app/controllers/careers/listing/index.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+        answeredElements = this.applicationStatus ? this.applicationStatus.responses.length : 0;
```
Comment (James1x0) [inline app/controllers/careers/listing/index.js]:
```suggestion
        answeredElements = ((this.applicationStatus || {}).responses || []).length;
```

File: app/templates/careers/listing/index.hbs [hunk 4]
```diff
@@ -1,4 +1,4 @@
-{{#if applicationStatus}}
+{{#if answered }}
```
Comment (James1x0) [inline app/templates/careers/listing/index.hbs]:
```suggestion
{{#if answered}}
```

File: app/templates/careers/listing/index.hbs [hunk 5]
```diff
@@ -13,9 +13,14 @@
 <h2 class="header">Description</h2>
 {{{model.description}}}
 
-{{#unless applicationStatus}}
+{{#if (and (not answered)previousApp)}}
```
Comment (James1x0) [inline app/templates/careers/listing/index.hbs]:
```suggestion
{{#if (and previousApp (not answered))}}
```

File: app/templates/careers/listing/index.hbs [hunk 6]
```diff
@@ -13,9 +13,14 @@
...
+  <div class="ui divider"></div>
+  <h2 class="ui center aligned header careers-listing__apply-btn">
+    {{#link-to 'careers.listing.apply.screen' previousApp class="text-green"}}Resume Application{{/link-to}}
+  </h2>
+{{else if (and (not answered previousApp))}}
```
Comment (James1x0) [inline app/templates/careers/listing/index.hbs]: Is a logic check here necessary?

**Comment by JavaLavaMT** [comment]
image

File: app/routes/careers/listing/index.js [hunk 7]
```diff
@@ -1,20 +1,26 @@
...
-    });
+    let previousApp = user ? await this.store.queryRecord('jobApplication', {
+      employee:   user,
+      jobOpening: jobOpening.id
+    }) : {};
+
+    return {
+      previousApp: previousApp ? previousApp.id : null,
+      listing:     jobOpening,
+      status:      await this.lookupJobApp(user, jobOpening.get('id'))
```
Comment (James1x0) [inline app/routes/careers/listing/index.js]: I'm not sure we need previousApp, isn't this the same as status??

File: package.json [hunk 8]
```diff
@@ -101,6 +101,7 @@
     "@bower_components/localforage": "mozilla/localForage#~1.3.1",
     "@bower_components/nprogress": "alexlafroscia/nprogress#allow-alternate-scheduler",
     "@bower_components/semantic-ui-calendar": "mdehoog/semantic-ui-calendar#^0.0.6",
-    "ember-cli-shims": "^1.2.0"
+    "ember-cli-shims": "^1.2.0",
+    "ember-concurrency": "^0.8.27"
```
Comment (James1x0) [inline package.json]: I believe I stripped the one concurrency call. Do you need this dependency anymore?

File: app/routes/careers.js [hunk 9]
```diff
@@ -1,31 +1,32 @@
 import Route from '@ember/routing/route';
```
Comment (James1x0) [inline app/routes/careers.js]: Remove extra fetch of applications here...

File: app/components/list-item/job-opening-item.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+const JobOpeningItemComponent = Component.extend({
+  tagName: '',
+
+  hasApplied: computed('applications.[]', 'job', function () {
+    const job = this.job;
+    return (this.jobApplications || []).find(app => {
+      return app.belongsTo('jobOpening').id() === job.id;
+    });
+  })
```
Comment (James1x0) [inline app/components/list-item/job-opening-item.js]: Remove this CP

File: app/components/list-item/job-opening-item.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+});
+
+JobOpeningItemComponent.reopenClass({ positionalParams: [ 'job', 'jobApplications' ] });
```
Comment (James1x0) [inline app/components/list-item/job-opening-item.js]: Remove jobApplications param

File: app/templates/components/list-item/job-opening-item.hbs [hunk 12]
```diff
@@ -0,0 +1,19 @@
+{{#link-to 'careers.listing' job.slug class="item careers__job-item" role="listitem"}}
+  <div class="ui header job-item__header">
+    <div class="content job-item__header-content text-blue clearfix">
+      {{job.title}}
+
+      <div class="sub meta header">
+        {{job.jobType}} -
+        {{job.location.addressCity}}, {{job.location.addressState}}
+        <div class="right floated">
+          {{#if hasApplied.isCompleted}}
```
Comment (James1x0) [inline app/templates/components/list-item/job-opening-item.hbs]: ❌

File: app/templates/careers/index.hbs [hunk 13]
```diff
@@ -1,36 +1,23 @@
...
-                    {{job.jobType}} -
-                    {{job.location.addressCity}}, {{job.location.addressState}}
-                    <div class="right floated">
-                      {{time (or job.startOn job.created)}}
-                    </div>
-                  </div>
-                </div>
-              </div>
-            {{/link-to}}
+            {{list-item/job-opening-item job model.jobApplications}}
```
Comment (James1x0) [inline app/templates/careers/index.hbs]:
```suggestion
            {{list-item/job-opening-item job}}
```

File: app/styles/quartzstyles/page/careers.less [hunk 14]
```diff
@@ -50,3 +48,7 @@
   margin-bottom: 5em;
   font-size: 2em;
 }
+
+.careers__job-category {
+  padding-top: 3em;
```
Comment (James1x0) [inline app/styles/quartzstyles/page/careers.less]:
```suggestion
  margin-bottom: 3em;
```

File: app/styles/quartzstyles/page/careers.less [hunk 15]
```diff
@@ -4,10 +4,7 @@
   margin-bottom: 2em;
   min-height: 45vh;
   box-shadow: 0 -4px 12px fade(@granite-grey, 75%) inset;
-  background-image: radial-gradient(
-    ellipse farthest-corner at 45px 45px,
-    darken(@granite-blue, 25%) 50%,
-    darken(@granite-green, 15%) 95%);
+  background-image: radial-gradient(ellipse farthest-corner at 45px 45px, darken(@granite-blue, 25%) 50%, darken(@granite-green, 15%) 95%);
```
Comment (James1x0) [inline app/styles/quartzstyles/page/careers.less]:
```suggestion
  background-image: radial-gradient(
    ellipse farthest-corner at 45px 45px,
    darken(@granite-blue, 25%) 50%,
    darken(@granite-green, 15%) 95%);
```

File: app/routes/careers.js (hunk 9 above)
Comment (JavaLavaMT) [reply app/routes/careers.js]: 🏁

File: app/components/list-item/job-opening-item.js (hunk 10 above)
Comment (JavaLavaMT) [reply app/components/list-item/job-opening-item.js]: 🏁

File: app/components/list-item/job-opening-item.js (hunk 11 above)
Comment (JavaLavaMT) [reply app/components/list-item/job-opening-item.js]: 🏁

File: app/routes/careers.js [hunk 16]
```diff
@@ -7,17 +7,14 @@ export default Route.extend({
...
     let jobQuery = {
       company: company.id,
       hiring:  true,
       sort:    { created: -1 }
     };
-
     if (!resolvedFromURL) {
       jobQuery.availableInternally = true;
     }
-
```
Comment (James1x0) [inline app/routes/careers.js]: Spacing in this file was intentional

File: app/templates/careers/index.hbs [hunk 17]
```diff
@@ -1,39 +1,23 @@
-{{#display/careers-hero palette=company.rgbPalette}}
-  <div class="company-logo">
-    <div class="company-logo__logo" style={{logoUrl}}></div>
-  </div>
-{{/display/careers-hero}}
-
-<main class="ui container" role="list">
-  <h1 class="ui centered aligned header">
```
Comment (James1x0) [inline app/templates/careers/index.hbs]: I think this change is unintentional

File: app/templates/careers/listing/apply/index.hbs [hunk 18]
```diff
@@ -53,7 +53,7 @@
         type="submit"
         class="ui huge fluid green button"
         loading=working
-        disabled=(not (or model.applicant.validations.isValid auth.user))
+        disabled=(and (not model.applicant.validations.isValid) (not this.auth.userId))
```
Comment (James1x0) [inline app/templates/careers/listing/apply/index.hbs]: I'm struggling to see the difference in this change other than this.auth*.userId*

File: app/templates/careers/index.hbs [hunk 19]
```diff
@@ -5,35 +5,14 @@
...
 
 <main class="ui container" role="list">
-  <h1 class="ui centered aligned header">
-    Careers at
-    {{#if company}}
-      {{company.name}}
-    {{else}}
-      Company
-    {{/if}}
-  </h1>
```
Comment (James1x0) [inline app/templates/careers/index.hbs]: This is marked as resolved, but still removing? Please *remove* this deletion. ;)

Comment (JavaLavaMT) [reply app/templates/careers/index.hbs]: yep not sure why i did that haha

**Pull request merged**